In [4]:
import pandas as pd
import numpy as np

In [5]:
file_path = "../data/raw/online_retail_II.csv"
df = pd.read_csv(file_path)
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [6]:
print("Rows:", df.shape[0])
print("Column:", df.shape[1])

Rows: 1067371
Column: 8


In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 65.1 MB


## 1. Duplicate Check

Duplicate transaction records can inflate sales, quantity, and customer mertrics.
Therefore, excat duplicate rows are checked before further analysis.

In [8]:
# Count exact duplicate rows
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 34335


In [9]:
# Preview duplicate rows
df[df.duplicated()].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom


In [12]:
# Dataset size
print("Total rows:", len(df))
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate percentage:", round(df.duplicated().mean() * 100, 2), "%")

Total rows: 1067371
Duplicate rows: 34335
Duplicate percentage: 3.22 %


In [13]:
df[df.duplicated(keep=False)].sort_values(
    by=["Invoice", "StockCode"]
).head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


In [15]:
duplicate = df[df.duplicated()]
duplicate.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
390,489517,84951A,S/4 PISTACHIO LOVEBIRD COASTERS,1,2009-12-01 11:34:00,2.55,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
657,489529,22028,PENNY FARTHING BIRTHDAY CARD,12,2009-12-01 11:51:00,0.42,17984.0,United Kingdom
658,489529,22036,DINOSAUR BIRTHDAY CARD,12,2009-12-01 11:51:00,0.42,17984.0,United Kingdom


In [18]:
df[df.duplicated()].iloc[0]

Invoice                          489517
StockCode                         21912
Description    VINTAGE SNAKES & LADDERS
Quantity                              1
InvoiceDate         2009-12-01 11:34:00
Price                              3.75
Customer ID                     16329.0
Country                  United Kingdom
Name: 371, dtype: object

In [19]:
df.iloc[379]

Invoice                                 489517
StockCode                                21491
Description    SET OF THREE VINTAGE GIFT WRAPS
Quantity                                     1
InvoiceDate                2009-12-01 11:34:00
Price                                     1.95
Customer ID                            16329.0
Country                         United Kingdom
Name: 379, dtype: object

In [20]:
# Remove excat duplicate rows
df = df.drop_duplicates()

# Verify duplicate removal
print("Remaining duplicate rows:", df.duplicated().sum())
print("Current dataset shape:", df.shape)

Remaining duplicate rows: 0
Current dataset shape: (1033036, 8)


### Duplicate Removal

A total of 34,355 excat duplicate rows were identified and removed.

Since all duplicated records contained identical values across every column, they were treated as redundant records rather than legitimate repeated purchases.

## 2. Missing Value Assessment 
Missing Value were assessed to dertermine whether they should be removed, retained, or imputed based on their impact on the analysis.

In [21]:
# Missing values by column
missing_values = df.isnull().sum()

missing_values

Invoice             0
StockCode           0
Description      4275
Quantity            0
InvoiceDate         0
Price               0
Customer ID    235151
Country             0
dtype: int64

In [22]:
# Missing value percentage
missing_percentage = (
    df.isnull().sum() / len(df) * 100
).round(2)

missing_percentage

Invoice         0.00
StockCode       0.00
Description     0.41
Quantity        0.00
InvoiceDate     0.00
Price           0.00
Customer ID    22.76
Country         0.00
dtype: float64

In [23]:
# Missing values summary

missing_summary= pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "missing_percentage": (
        df.isnull().sum() / len(df) * 100
    ).round(2)
})

missing_summary

,Missing Count,missing_percentage
Invoice,0,0.00
StockCode,0,0.00
Description,4275,0.41
Quantity,0,0.00
InvoiceDate,0,0.00
Price,0,0.00
Customer ID,235151,22.76
Country,0,0.00


In [27]:
# Rows with missing Description
df[df["Description"].isnull()].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,NaN,United Kingdom
6378,489882,35751C,NaN,12,2009-12-02 16:22:00,0.0,NaN,United Kingdom
6555,489898,79323G,NaN,954,2009-12-03 09:40:00,0.0,NaN,United Kingdom
6576,489901,21098,NaN,-200,2009-12-03 09:47:00,0.0,NaN,United Kingdom
6581,489903,21166,NaN,48,2009-12-03 09:57:00,0.0,NaN,United Kingdom


In [28]:
df[df["Description"].isnull()].shape

(4275, 8)

In [30]:
# Check missing descriptions with Zero price
df[
   (df["Description"].isnull()) &
   (df["Price"] ==0)
   ].shape

(4275, 8)

In [31]:
 # Summary statistics of missing description rows
df[df["Description"].isnull()].describe(include="all")

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
count,4275,4275,0,4275.000000,4275,4275.0,0.0,4275
unique,4275,2451,0,NaN,2999,NaN,NaN,1
top,489521,84990,NaN,NaN,2010-10-11 10:24:00,NaN,NaN,United Kingdom
freq,1,12,NaN,NaN,6,NaN,NaN,4275
mean,NaN,NaN,NaN,-17.366784,NaN,0.0,NaN,NaN
std,NaN,NaN,NaN,515.938734,NaN,0.0,NaN,NaN
min,NaN,NaN,NaN,-9600.000000,NaN,0.0,NaN,NaN
25%,NaN,NaN,NaN,-28.000000,NaN,0.0,NaN,NaN
50%,NaN,NaN,NaN,-4.000000,NaN,0.0,NaN,NaN
75%,NaN,NaN,NaN,3.000000,NaN,0.0,NaN,NaN


### Handling Missing Values

The 'Description' column contained 4,275 missing values (0.41% of the dataset). 

Further investigation showed that these records also had a unit price of 0 and missing Customer IDs, indicating imcomplete or invalid transaction records rather than legitimate sales.

Therefore, rows with missing product descriptions were removed.

Missing 'Customer ID' values were retained because they are only required for customer-level analyses such as RFM segmentation and are not necessary for sales or product analysis.


In [32]:
# Remove Rows with missing Description
df = df.dropna(subset=["Description"])

In [33]:
print(df.isnull().sum())
print(df.shape)

Invoice             0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
Price               0
Customer ID    230876
Country             0
dtype: int64
(1028761, 8)


In [35]:
rows_before = len(df)

df= df.dropna(subset=["Description"])

rows_after = len(df)

print(f"Rows before cleaning : {rows_before:,}")
print(f"Rows after cleaning : {rows_after:,}")
print(f"Rows removed         : {rows_before - rows_after:,}")

Rows before cleaning : 1,028,761
Rows after cleaning : 1,028,761
Rows removed         : 0


In [36]:
df["Description"].isnull().sum()

np.int64(0)

In [37]:
print(df.shape)

(1028761, 8)


## 3. Quantity Validation

The Quantity column was examined to identify invalid or unusual transaction quantities.

Negative quantities were investigated because they may represent returned or cancelled orders rather than data quality issues. 

In [38]:
df["Quantity"].describe()

count    1.028761e+06
mean     1.019092e+01
std      1.723738e+02
min     -8.099500e+04
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

In [40]:
negative_qty = df[df["Quantity"] < 0]
print("Negative Quantity Records:", len(negative_qty))

Negative Quantity Records: 19863


In [41]:
negative_qty.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


## 3. Quantity Validation

The 'Quantity' column was examined to identify zero, negative, and unusually large values.

Negative quantities may represent product returns, cacellations, or inventory adjustments rather than data entry errors.

In [42]:
# Summary statistics for Quantity
df["Quantity"].describe()

count    1.028761e+06
mean     1.019092e+01
std      1.723738e+02
min     -8.099500e+04
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

In [44]:
# Count zero and negative quantity records
zero_qty_count = (df["Quantity"]==0).sum()
negative_qty_count = (df["Quantity"]<0).sum()
positive_qty_count = (df["Quantity"]>0).sum()

print(f"Zero_quantity records    : {zero_qty_count:,}")
print(f"Negative_quantity records: {negative_qty_count:,}")
print(f"Positive quantity records: {positive_qty_count:,}")

Zero_quantity records    : 0
Negative_quantity records: 19,863
Positive quantity records: 1,008,898


In [45]:
# Preview negative quantity records
negative_qty = df[df["Quantity"] <0]

negative_qty.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


### Interpretation

A total of 19,863 records contained negative quantities.

Further investigation showed that these transactions had invoice numbers beginning with the letter **C**, indicating cancelled or returned orders.

However, a small number of negative quantity records did not follow this pattern. These records will be investigated separately before making any cleaning decisions. 

At this stage, no negative quantity records are removed.

In [46]:
# Verify whether all negative quantities are cancellation invoices
negative_qty["Invoice"].astype(str).str.startswith("C").value_counts()

Invoice
True     19103
False      760
Name: count, dtype: int64

In [47]:
# Investigate negative quantities without cancellation invoice
negative_without_c = negative_qty[
    ~negative_qty["Invoice"].astype(str).str.startswith("C")
]

negative_without_c.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,NaN,United Kingdom
6556,489899,79323GR,sold as gold,-954,2009-12-03 09:41:00,0.0,NaN,United Kingdom
6911,490007,84347,21494,-720,2009-12-03 12:09:00,0.0,NaN,United Kingdom
9308,490130,21493,lost?,-600,2009-12-03 18:28:00,0.0,NaN,United Kingdom
17427,490765,21450,damaged,-31,2009-12-08 10:59:00,0.0,NaN,United Kingdom


In [48]:
negative_without_c.shape

(760, 8)

### Business Interpretation

Negative quantity records were devided into two categories:

- Customer return or calcellation transactions (Invoice begins with **C**)
- Interanl inventory adjustment records (e.g., *damaged*, *lost*, *short*, *bad quality*)

Return transactions were retained becaue they represent valid business events.

Inventory adjustment records were removed because they do not represent customer purchases and are outside the scope of this sales analysis project.

In [49]:
inventory_adjustment = negative_without_c

print("Inventory_adjustment records:", len(inventory_adjustment))

Inventory_adjustment records: 760


In [50]:
df = df.drop(inventory_adjustment.index)

In [52]:
print(df.shape)
print(
    df[
        (df["Quantity"] < 0) &
        (~df["Invoice"].astype(str).str.startswith("C"))
    ].shape
)

(1028001, 8)
(0, 8)


In [54]:
inventory_adjustment = negative_without_c

#Save for audit
inventory_adjustment.to_csv(
    "../data/processed/inventory_adjustments.csv",
    index=False
)

## 4. Price Validation

The 'Price' column was examined to identify zero or negative values.

Rather than assuming these records were invalid, zero-priced transactions were investigated to determine whether they represented legetimate business events or data quality issues.

In [55]:
# Summary statistics for Price
df["Price"].describe()

count    1.028001e+06
mean     4.636578e+00
std      1.226965e+02
min     -5.359436e+04
25%      1.250000e+00
50%      2.100000e+00
75%      4.150000e+00
max      3.897000e+04
Name: Price, dtype: float64

In [56]:
zero_price = df[df["Price"] == 0]
print(f"Zero price records: {len(zero_price):,}")

Zero price records: 979


In [57]:
zero_price.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126.0,United Kingdom
5904,489861,DOT,DOTCOM POSTAGE,1,2009-12-02 14:50:00,0.0,NaN,United Kingdom
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658.0,United Kingdom
16107,490727,M,Manual,1,2009-12-07 16:38:00,0.0,17231.0,United Kingdom
18738,490961,22065,CHRISTMAS PUDDING TRINKET POT,1,2009-12-08 15:25:00,0.0,14108.0,United Kingdom
18739,490961,22142,CHRISTMAS CRAFT WHITE FAIRY,12,2009-12-08 15:25:00,0.0,14108.0,United Kingdom
31993,491971,85042,ANTIQUE LILY FAIRY LIGHTS,1,2009-12-14 18:37:00,0.0,NaN,United Kingdom
32916,492079,85042,ANTIQUE LILY FAIRY LIGHTS,8,2009-12-15 13:49:00,0.0,15070.0,United Kingdom
40101,492760,21143,ANTIQUE GLASS HEART DECORATION,12,2009-12-18 14:22:00,0.0,18071.0,United Kingdom
47126,493761,79320,FLAMINGO LIGHTS,24,2010-01-06 14:54:00,0.0,14258.0,United Kingdom


## 5. Invoice Validation

Invoice numbers were checked to distinguish completed sales from cancelled transactions.

Invoices beginning with **C** represent cancelled or returned orders and were retained for future return analysis.

In [59]:
cancelled_orders = df[
    df["Invoice"].astype(str).str.startswith("C")
]

print(f"Cancelled Orders: {len(cancelled_orders):,}")

Cancelled Orders: 19,104


In [60]:
cancelled_orders.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


In [62]:
df.to_csv(
    "../data/processed/online_retail_clean.csv",
    index= False
    )

print("Clean dataset saved successfully.")

Clean dataset saved successfully.


In [64]:
from pathlib import Path

Path("../data/processed/online_retail_clean.csv").exists()

True

# Summary of Data Cleaning
## Cleaning Tasks Completed
✔ Removed exact duplicate records
✔ Removed incomplete records with missing product descriptions.
✔ Retained customer returned transactions for future business analysis.
✔ Removed internal inventory adjustment records (damaged, lost, bad quality, etc.).
✔ Validated quantity values.
✔ Validated price values.
✔ Validated invoice numbers.

## Business Rules Applied

- Exact duplicate records were removed.
- Missing product descriptions were treated as incomplete transactions.
- Customer return transactions were preserved.
- Internal inventory adjustment records were excluded from Customer sales analysis.

## Result 

The cleanded dataset is now ready for:

- Explaoratory Data Analysis (EDA)
- Sales Trend Analysis 
- Product Performance Anlysis
- Customer Segmentation (RFM)